src/adapter/src/rbac.rs

```diff
@@ -611,47 +633,93 @@ fn generate_required_privileges(
             let mut privileges =
                 generate_read_privileges(catalog, depends_on.iter().cloned(), role_id);
             if let Ok(cluster) = catalog.resolve_cluster(None) {
-                privileges.insert((cluster.id().into(), AclMode::USAGE, role_id));
+                privileges.push((cluster.id().into(), AclMode::USAGE, role_id));
             }
             privileges
         }
         Plan::Explain(_) => generate_read_privileges(catalog, depends_on.iter().cloned(), role_id),
-        Plan::Insert(plan) => {
+        Plan::CopyFrom(plan) => {
             let item = catalog.get_item(&plan.id);
-            BTreeSet::from([
+            vec![
                 (
                     item.name().qualifiers.clone().into(),
                     AclMode::USAGE,
                     role_id,
                 ),
                 (plan.id.into(), AclMode::INSERT, role_id),
-            ])
+            ]
         }
-        Plan::CopyFrom(plan) => {
-            let item = catalog.get_item(&plan.id);
-            BTreeSet::from([
-                (
-                    item.name().qualifiers.clone().into(),
-                    AclMode::USAGE,
-                    role_id,
-                ),
+        Plan::Insert(plan) => {
+            let update_schema_id: ObjectId =
+                catalog.get_item(&plan.id).name().qualifiers.clone().into();
+            let mut privileges = vec![
+                (update_schema_id.clone(), AclMode::USAGE, role_id),
                 (plan.id.into(), AclMode::INSERT, role_id),
-            ])
+            ];
+            // plan.values has not been optimized yet, so it might be constant even if
+            // `plan.values.as_const().is_none()` is true.
+            privileges.extend_from_slice(&generate_read_privileges_inner(
```

Comment:
Do we need methods that do this plan walking in general so we don't miss adding priv requirements either now or in the future? This current structure seems like a thing that other people will have to know about when they add new SQL functionality because the compiler won't complain at them. A walk or visit method should either be auto generated or else use match blocks that will complain if new sql stuff isn't added to the new structs.

Reply:
I thought about something like this when implementing the feature. I'm just not sure exactly what the plan walking would look like since you need some context at each node. For example if we encounter a schema, then it could either be `USAGE` or `CREATE` depending on the context of that schema. Also I don't think we have the machinery to do a full plan walk right now.

One thought I had was to fully expand all plan fields in this method, so that if someone adds a field the compiler will complain.

Reply:
That feels useful even though it adds some annoyance, but that's kinda RBAC in general. I'm in support of that.

Reply:
Done